pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisConsumer.java

```diff
@@ -0,0 +1,194 @@
+    super(kinesisConfig.getStream(), kinesisConfig.getAwsRegion());
+    _stream = kinesisConfig.getStream();
+    _maxRecords = kinesisConfig.maxRecordsToFetch();
+    _shardIteratorType = kinesisConfig.getShardIteratorType();
+    _executorService = Executors.newSingleThreadExecutor();
+  }
+
+  public KinesisConsumer(KinesisConfig kinesisConfig, KinesisClient kinesisClient) {
+    super(kinesisConfig.getStream(), kinesisConfig.getAwsRegion(), kinesisClient);
+    _kinesisClient = kinesisClient;
+    _stream = kinesisConfig.getStream();
+    _maxRecords = kinesisConfig.maxRecordsToFetch();
+    _shardIteratorType = kinesisConfig.getShardIteratorType();
+    _executorService = Executors.newSingleThreadExecutor();
+  }
+
+  /**
+   * Fetch records from the Kinesis stream between the start and end KinesisCheckpoint
+   */
+  @Override
+  public KinesisRecordsBatch fetchMessages(StreamPartitionMsgOffset startCheckpoint,
+      StreamPartitionMsgOffset endCheckpoint, int timeoutMs) {
+    List<Record> recordList = new ArrayList<>();
+    Future<KinesisRecordsBatch> kinesisFetchResultFuture =
+        _executorService.submit(() -> getResult(startCheckpoint, endCheckpoint, recordList));
+
+    try {
+      return kinesisFetchResultFuture.get(timeoutMs, TimeUnit.MILLISECONDS);
+    } catch (Exception e) {
+      return handleException((KinesisPartitionGroupOffset) startCheckpoint, recordList);
+    }
+  }
+
+  private KinesisRecordsBatch getResult(StreamPartitionMsgOffset start, StreamPartitionMsgOffset end,
+      List<Record> recordList) {
+    KinesisPartitionGroupOffset kinesisStartCheckpoint = (KinesisPartitionGroupOffset) start;
+
+    try {
+
+      if (_kinesisClient == null) {
+        createConnection();
+      }
+
+      //TODO: iterate upon all the shardIds in the map
+      // Okay for now, since we have assumed that every partition group contains a single shard
+      Map.Entry<String, String> shardToSequenceNum =
+          kinesisStartCheckpoint.getShardToStartSequenceMap().entrySet().iterator().next();
+      String shardIterator = getShardIterator(shardToSequenceNum.getKey(), shardToSequenceNum.getValue());
+
+      String kinesisEndSequenceNumber = null;
+
+      if (end != null) {
+        KinesisPartitionGroupOffset kinesisEndCheckpoint = (KinesisPartitionGroupOffset) end;
+        kinesisEndSequenceNumber = kinesisEndCheckpoint.getShardToStartSequenceMap().values().iterator().next();
+      }
+
+      String nextStartSequenceNumber = null;
+      boolean isEndOfShard = false;
+
+      while (shardIterator != null) {
+        GetRecordsRequest getRecordsRequest = GetRecordsRequest.builder().shardIterator(shardIterator).build();
+        GetRecordsResponse getRecordsResponse = _kinesisClient.getRecords(getRecordsRequest);
+
+        if (getRecordsResponse.records().size() > 0) {
+          recordList.addAll(getRecordsResponse.records());
+          nextStartSequenceNumber = recordList.get(recordList.size() - 1).sequenceNumber();
```

Comment:
If we are iterating through shards then we should be getting a vector of sequence numbers, right? We seem to be keeping track of the sequence number of the last shard we examined?
Othewise add a PreCondition that the shardlist has exactly one shard.

Reply:
Added the precondition. Right now we have made the assumption that we will only have 1 shard per segment